### ISSUE TITLE
PROCESS CHANGE: Overhaul and standardize Knative roadmap processes

### ISSUE THREAD

**Issue by omerbensaadon** [knative/community#746]
## Background
Ahmed and I have been collaborating to improve roadmapping practices in the Knative Community. 

These changes ought to be formalized at TOC "WG Updates" in order to create healthier processes within WGs, help new contributors understand where they can chip in and liberate us from the [tyranny of time-bound objectives.](https://memegenerator.net/img/instances/73848759/the-gentle-laborer-shall-no-longer-suffer.jpg)

## Considerations + Prior Art
There are some Working Groups in the Knative community (e.g. DUX WG, Security WG) where contributor priorities are mostly the aligned, and are set by some sort of collective agreement or collaboration. In these cases, **Priority** is important to convey to contributors.

There are still others (e.g. Eventing WG) where the contributor priorities are not all aligned, and efforts are focused on numerous vendor-specific parallel efforts. In these cases, **Alignment** is important to convey to fellow contributors.
```
Priority = an understanding of what work needs to be done first, second, etc.
Alignment = an understanding of what work is being executed on, will be executed and by whom. 
```

The needs between these two WGs is different, but the way they present which work is being done and when should not differ.

### Feature Track Documents
While Feature Track Documents are useful for expressing complex ideas and discussing the details of proposals, **it's hard to argue that they have been an effective means of tracking work.** 

### DUX WG
Miro for Site Revamp: https://miro.com/app/board/o9J_lGNC-LY=/ 
Milestone Planning: https://github.com/knative/docs/milestone/23
UX WG Highlevel Roadmap: https://github.com/orgs/knative/projects/20
### Eventing WG
Roadmap Markdown: https://github.com/knative/eventing/blob/main/docs/roadmap-2021.md

Eventing 1.0 Project: https://github.com/knative/eventing/projects/19

---

# Process Changes
## (NEW) Move Roadmaps to Github Projects
Currently, the WG Roadmap can range anywhere from a .md file to a PPT slide. Across WGs, there is not a standardized format or location for the work being undertaken. This presents a barrier to entry for new contributors or vendors who may want to understand (a) where they can chip in or (b) what work has been prioritized.

**We do all of our work in GH, our roadmaps should be no different.** The proposal here is to mandate that groups come to the ToC with roadmaps in [Knative's Org](https://github.com/orgs/knative/projects), with each item listed as an Issue, and strongly recommend they conform to the "Lightweight Guidance" described below. 

## (NEW) Remove Timelines from TOC Roadmap Updates
Having roadmaps tied to amounts of time is either (a) not taken at-face-value by community members ("we just mean we'll get to that soon") or, worse yet, (b) is actively misleading. 

It also promotes a "plan everything up front" processes where members of Working Groups need to come together and make WAGs (Wild-Ass Guesses) about how much work they will get done in the coming months.

My proposal is to replace "Next 3 months" with `Next`, "Next 6 months" with `Later`, add the columns `Now` and `Icebox` as well  as instituting lightweight guidance for using these columns.

#### `In Progress` Column
This list feeds into `What's Landed` and is replenished by `Ready To Work`.

Lightweight Guidance:
- **Work in this column ought to be fleshed out enough to promote contribution.**
- This column ought to be capped at to approximately 1-2 issues per contributor.
- Each issue in this column has an owner.
- Each issue in this column links to something that organizes efforts (more on this below). 

#### `Ready To Work` Column
 This list feeds into `Now` and is replenished by `In Design`.

Lightweight Guidance:
- ==Work in this column ought to be fleshed out enough to promote implementation, but does not have someone currently working on it.==
- Capped to ~1.5 times the number of items in `Now`.
- Each issue in this column links to something that organizes efforts (issues, other GH Project boards, Milestones, etc).

#### `In Design` Column
The efforts listed in the `Later` column ought to be efforts that the WG agrees will be worked at some point in the future, without a firm understanding of when that will be. 

This list feeds into `Ready To Work` and is replenished by the `Icebox`.

Lightweight Guidance:
- **Work in this column ought to be fleshed out enough to promote discussion.**
- May need a Feature Track or other planning document.
- Capped to ~1.5 times the number of items in `Next`.
- Each issue in this column links to something that organizes efforts (issues, other GH Project boards, Milestones, etc).

#### `Icebox` Column
The Icebox is a totally unstructured list where anyone can throw any idea they would like, leaving up to the individual WGs what "unstructured" means to them.

Lightweight Guidance:
- **Work in this column does not need to be fleshed out.**
- Capped to ~1.5 times the number of items in `Next`.

### Moving Items Between Groupings 
Ideally, the columns of this roadmap should be 1-in-1-out. The WG Leads should take the responsibility to move items between these columns. 

* Expected benefits. Who gains the benefits? Why will they benefit?
Described above
* Expected costs. Who bears the costs? How heavy are they?
There are expected costs for WG Leads, but ideally this makes their jobs a bit easier. 
* Timeframe for implementation / rollout.
ASAP
* Are you willing to drive the process, or is this a request for help?
Of course!

**Comment by omerbensaadon** [comment]
If this issue is accepted, follow-on issues will be opened for the following: 

1. Creating a Roadmap item template 
2. Establishing Quasi-Leadership positions for Roadmap Item Owners 
3. Having each Roadmap item link to its own Project board

**Comment by evankanderson** [comment]
Quick thoughts before the meeting:

# Project board vs Markdown

## Pros of project boards

* Issues align naturally with owners, etc
* Encourages documentation within an effort by using the issue as a log of work planning and effort
* Issues are at least somewhat discoverable to new contributors, though the project board might be missed at first
* Org level project boards are open to manipulation by all org members (easier maintenance)

## Pros of Markdown

* Long form prose makes it easier to document dependencies and rationale for ordering
* Roadmap as a file in a repo is probably the most discoverable for new contributors
* Markdown files require two-phase commit (pr + approval), making it harder for a lone actor to subvert group consensus

I could be convinced that project boards are the way to go.

# Naming and item limits

While it seems reasonable practice to cap the amount of work in progress, I'm not sure we need specific guidance about the number of items in each column / category. I'm also wondering if some naming changes would help improve comprehension and align with GitHub automation:

* Now --> In Flight / In Progress (move here when assigned)
* Next --> Ready to Work (moved manually by WG leads when it meets the implementation bar)
* Later --> In Design (not sure if we'd use assignment from Icebox here or a manual process)
* Icebox seems fine, or "Future Work" as a holding box for known gaps not currently being addressed

# Visibility, Discoverablity, and Process

In addition to the project boards, we need some cross-references so that new contributors can find the board(s). My suggestion:

* Convert the policy in the issue to a file in `mechanics`
* Add a link for each Working Group to the index in community
* For each repo, add a `ROADMAP.md` that links to the policy and has a heading for each WG owning files in that repo with a link to their roadmap project. e.g.
  ```md
  For details on the roadmap process [see the common Knative policy] (https://github.com/knative/community/tree/main/mechanics/ROADMAPS.md)

  # Eventing WG
  [This project](https://github.com/orgs/knative/projects/44)

  # Sources WG
  [This project](https://github.com/orgs/knative/projects/47)
  ```

**Comment by evankanderson** [comment]
+1 from Evan

**Comment by rhuss** [comment]
Thanks a lot for the initiative, I also believe that the current time-based format to plan/present roadmaps does not reflect the reality.

I'm +1 in general, with some comments:

* +1 for using GitHub Projects over Markdown as planning should be with low friction and very little process to change (markdown and PR process does not replace the mutual trust that is required anyway, so I don't see subverting group consensus as an issue at all). I agree we should refer to the projects from various entry places (repo README's, kative.dev mostly)
* I also won't enforce a fixed WIP limit that is the same for all repositories since the number of people working in the different areas vary. So, I wouldn't enforce a WIP limit of 3 on the "Now" column, but either don't work with WIP limits or let the individual working group decide what to chose.
* It should become a natural part of the WG call to look to this board. Maybe not every week, but at least once a month or so. This might be also documented in the community guidelines on knative.de
* +1 for not using time-based columns (like the client started in a half-baked balloon to match the TOC update format --> https://github.com/knative/client/projects/4). One question would be whether we should map the columns to forthcoming Knative releases. I think this might be helpful to manage expectations, but I also understand that this should not be taken as a guarantee. But maybe we could map "Now" to "the next Knative release" ? and "Next" to "the over next or over-over next release") ?
* "Icebox" is kind of an unknown metaphor for me (sorry for my ignorance ;-). What's about using "Backlog" ? (but that might be not well anticipated by others). Also, I wouldn't expose a WIP limit on that column but should be cleaned up in a triage process periodically.

**Comment by markusthoemmes** [comment]
+1 on making the roadmaps be in Github Projects. -1 as @rhuss stated, on the WIP limits.

**Comment by omerbensaadon** [comment]
I like the idea of limiting a person to a # of Roadmap Items from @rhuss

**Comment by omerbensaadon** [comment]
I am going to drop WIP limits, TOC will be responsible for assessing people's roadmap

**Comment by evankanderson** [comment]
It's worth including some non-rules guidance about how WGs should look at execution of "in flight" items.

**Comment by evankanderson** [comment]
Approved at 16 Sep 2021 TOC meeting

**Comment by dprotaso** [comment]
After going through the process of setting up a board a few things I noticed/we should recommend

1. Make sure folks are creating boards using the beta projects feature at the org leve
   - https://github.com/orgs/knative/projects?type=beta
1. Make sure the board is set to public
   - A bunch of boards are private
1. Make sure columns names are consistently named and ordered


I also think we should consider dropping the 'Icebox' - since I think that'll be noisy and it doesn't convey anything useful. A roadmap is what _we are planning_ on working on

The icebox can exist in a different location - ie. for serving we have a milestone for it https://github.com/knative/serving/milestone/17

**Comment by rhuss** [comment]
+1 to keep only those task in the roadmap that are in the focus for the next time and that can be actually worked on.


Actually, for planning, I'd love to have an extra column "Proposed" for the planning session. So people do preselection of features that they want to see on the roadmap prior to the roadmap session. Then, we discuss those and move some of them to the "In Design" column (we called it "Selected") where then start the consolidation of the issue so that we then can move them to "Read For Work" so people can pick those up, assign them to themselves and move it to "In Progress". We tried that the first time yesterday with the client board at https://github.com/orgs/knative/projects/37/views/1 which worked smoothly.

After the planning session, the "Proposed" column then is emptied again (or even removed).

Minor enhancement proposal: Reorder from "In Design" / "Selected" -> "Read to Work" -> "In Progress" from left to right since this seems the more natural flow that most people are used to when working with such boards (instead of from right to left as it is now).

**Comment by kvmware** [comment]
Capturing some points from TOC meeting that might already be expressed in the above but repeating for my sake here. Icebox is for items that don't have owners and/or designs and is more of a wishlist. The ready-to-work column is for work that has owners or has a flushed out design and would like to be picked up? All this once flushed out should be in a markdown file in community/mechanics repo.

**Comment by evankanderson** [comment]
The **Ready To Work** column is for:

> Work in this column ought to be fleshed out enough to promote implementation, but does not have someone currently working on it.

The idea is that there's a nebulous / precise feature request, but it may require some amount of planning or design (and consensus) that needs to be done before code starts to be committed. For a small, clear feature, it might flow from "Icebox" --> "Ready To Work" directly, whereas a large feature might spend a while on "In Design" with a Feature Track document being written / reviewed.

The intent is that "Ready To Work" items _do not_ have an owner -- these are opportunities for either new contributors, or contributors who've finished an existing track.

One interesting question is how to track work that has a built-in timeout, like https://github.com/knative/eventing/issues/5756. For this issue, the main code has been submitted, but there's a follow-on for each release to move the feature flag forward. I'm not sure whether creating additional issues like `Post-1.1: move $FEATURE to Beta` and putting those in "Ready-to-Work" would help in keeping Work-In-Progress bounded.

**Comment by evankanderson** [comment]
> Minor enhancement proposal: Reorder from "In Design" / "Selected" -> "Read to Work" -> "In Progress" from left to right since this seems the more natural flow that most people are used to when working with such boards (instead of from right to left as it is now).

This is interesting -- I think this depends on your frame of reference. If you're an end-user looking forward to "In Progress" items, then having things flow RTL into "stuff I'm paying attention to / using" may make sense. If you're a feature looking at "when do I get released", then there's a LTR flow through the stages of "collection", "design", "execution" (possibly sitting in the "Ready To Work" bin until someone picks it up for execution).
